### ISSUE TITLE
Extremely slow start of the OracleDatabase 19.3 se2 docker container

### ISSUE THREAD

**Issue by vmassol** [oracle/docker-images#1480]
Hi, I successfully built the docker image for OracleDatabase 19.3 se2.

The issue is when I start it. I've tried this on 2 different machines (on a Mac and one a Linux system) and in both cases it takes more than 35 minutes to start! 

I started it with `docker run -p 1521:1521 -p 5500:5500 -e ORACLE_SID=xwiki -e ORACLE_PDB=xwikipdb -e ORACLE_PWD=xwiki  oracle/database:19.3.0-se2` (note: I also tried by mapping a local directory for the DB files with the same slow result).

The part that is extremely slow is this part:

```
...
Prepare for db operation
8% complete
Copying database files
31% complete
Creating and starting Oracle instance
32% complete
36% complete
40% complete
43% complete
46% complete
Completing Database Creation
51% complete
54% complete
Creating Pluggable Databases
58% complete
77% complete
Executing Post Configuration Actions
100% complete
...
```

We used to use a Oracle 11g docker image execute test XWiki (https://xwiki.org) on it to validate that we support the Oracle DB and it was working ok (about 2mn to start it). We'd like to validate against a more recent version, which is why I've been trying using version 19.3 for the past 2 days.

However with a startup time of 35+ minutes it's impossible for us to use it for our tests. We start and stop the container hundreds of times across all our test suite (hundreds if not thousands of tests across different modules).

So some questions:
1) Is that startup expected? Is that what everyone is experiencing? I should note that the machines on which I tried it are quite large and beefy machines with lots of CPU and RAM (and I didn't put any resource limits).
2) Is there something that can be done to speed it up so that it starts in about 2-3 minutes?

Thanks a lot
-Vincent
XWiki open source commmitter

**Comment by ggajos** [comment]
I posted a blog post (instruction), why it is taking so much time and how it can be optimized here: https://medium.com/@ggajos/drop-db-startup-time-from-45-to-3-minutes-in-dockerized-oracle-19-3-0-552068593deb

**Comment by vmassol** [comment]
@ggajos Thanks a lot, very useful. I actually used some Oracle 19.3 images that were supposed to do this but they were still quite slow to start if I remember correctly: https://hub.docker.com/r/quillbuilduser/oracle-18-xe

3 minutes is still very long to start to DB (we have hundreds if not thousands of tests in different modules) so that a lot of hours just to start/stop the Oracle DB... ;) But not much we can do about this except drop testing on Oracle or test less often on it.

The main issue with your solution is that we need to store the new image somewhere so that all devs can get it, including our CI. So this means having to setup a Docker Registry, which we've avoided so far (lots of hassle and maintenance required). We're trying to use only dockerhub.

Thanks!

**Comment by KamilJedrzejuk** [comment]
If you don't have to use Oracle, don't use it. Better choice willbe a PostgreSQ.
Oracle is not the best choice for development. If your business logic is written in PLSQL you are out of luck. I was trying setup a docker image with Oracle 18, but now I see that it doesn'e make any sense, as Developer you can not use a nice library like Testcontainers…  @ogho-dockerhub

**Comment by KamilJedrzejuk** [comment]
It took about 10 minutes... Anu suggestion how to resolve this problem... 

$ docker run kamiljedrzejuk/oracle18c
Database does not exists, configuring
Configuring Oracle Listener.
Listener configuration succeeded.
Configuring Oracle Database XE.
Enter SYS user password:
********
Enter SYSTEM user password:
*******
Enter PDBADMIN User Password:
*********
Prepare for db operation
7% complete
Copying database files
29% complete
Creating and starting Oracle instance
30% complete
31% complete
34% complete
38% complete
41% complete
43% complete
Completing Database Creation
47% complete
50% complete
Creating Pluggable Databases
54% complete
71% complete
Executing Post Configuration Actions
93% complete
Running Custom Scripts
100% complete
Database creation complete. For details check the logfiles at:
 /opt/oracle/cfgtoollogs/dbca/XE.
Database Information:
Global Database Name:XE
System Identifier(SID):XE
Look at the log file "/opt/oracle/cfgtoollogs/dbca/XE/XE.log" for further details.

Connect to Oracle Database using one of the connect strings:
     Pluggable database: 054a9658b93b/XEPDB1
     Multitenant container database: 054a9658b93b
Use https://localhost:5500/em to access Oracle Enterprise Manager for Oracle Database XE
The Oracle base remains unchanged with value /opt/oracle
Enabling XDB for external access

SQL*Plus: Release 18.0.0.0.0 - Production on Sat Mar 14 13:03:16 2020
Version 18.4.0.0.0

Copyright (c) 1982, 2018, Oracle.  All rights reserved.


Connected to:
Oracle Database 18c Express Edition Release 18.0.0.0.0 - Production
Version 18.4.0.0.0

SQL>
PL/SQL procedure successfully completed.

SQL>
PL/SQL procedure successfully completed.

SQL> Disconnected from Oracle Database 18c Express Edition Release 18.0.0.0.0 - Production
Version 18.4.0.0.0
The following output is now a tail of the alert.log:
2020-03-14T13:03:13.272916+00:00
XEPDB1(3):Resize operation completed for file# 10, old size 358400K, new size 368640K
2020-03-14T13:03:14.919414+00:00
XEPDB1(3):CREATE SMALLFILE TABLESPACE "USERS" LOGGING  DATAFILE  '/opt/oracle/oradata/XE/XEPDB1/users01.dbf' SIZE 5M REUSE AUTOEXTEND ON NEXT  1280K MAXSIZE UNLIMITED  EXTENT MANAGEMENT LOCAL  SEGMENT SPACE MANAGEMENT  AUTO
XEPDB1(3):Completed: CREATE SMALLFILE TABLESPACE "USERS" LOGGING  DATAFILE  '/opt/oracle/oradata/XE/XEPDB1/users01.dbf' SIZE 5M REUSE AUTOEXTEND ON NEXT  1280K MAXSIZE UNLIMITED  EXTENT MANAGEMENT LOCAL  SEGMENT SPACE MANAGEMENT  AUTO
XEPDB1(3):ALTER DATABASE DEFAULT TABLESPACE "USERS"
XEPDB1(3):Completed: ALTER DATABASE DEFAULT TABLESPACE "USERS"
2020-03-14T13:03:16.148462+00:00
ALTER PLUGGABLE DATABASE XEPDB1 SAVE STATE
Completed: ALTER PLUGGABLE DATABASE XEPDB1 SAVE STATE

**Comment by KamilJedrzejuk** [comment]
After restart I have some errrors. BRAVO ORACLE …
docker run  -v oradata:/opt/oracle/oradata kamiljedrzejuk/oracle18c
Database exists
Starting Oracle Net Listener.
Oracle Net Listener started.
Starting Oracle Database instance XE.
Oracle Database instance XE started.

The following output is now a tail of the alert.log:
XEPDB1(3):Undo initialization online undo segments: err:0 start: 5365850 end: 5366119 diff: 269 ms (0.3 seconds)
XEPDB1(3):Undo initialization finished serial:0 start:5365847 end:5366124 diff:277 ms (0.3 seconds)
XEPDB1(3):Database Characterset for XEPDB1 is AL32UTF8
2020-03-14T13:36:49.591851+00:00
XEPDB1(3):Opening pdb with Resource Manager plan: DEFAULT_PLAN
Pluggable database XEPDB1 opened read write
Starting background process CJQ0
2020-03-14T13:36:50.189923+00:00
CJQ0 started with pid=44, OS id=315
Completed: ALTER DATABASE OPEN
2020-03-14T13:36:50.234450+00:00
Shared IO Pool defaulting to 16MB. Trying to get it from Buffer Cache for process 96.
===========================================================
Dumping current patch information
===========================================================
No patches have been applied
===========================================================
2020-03-14T13:37:01.451281+00:00
TABLE SYS.WRI$_OPTSTAT_HISTHEAD_HISTORY: ADDED INTERVAL PARTITION SYS_P221 (43903) VALUES LESS THAN (TO_DATE(' 2020-03-15 00:00:00', 'SYYYY-MM-DD HH24:MI:SS', 'NLS_CALENDAR=GREGORIAN'))
TABLE SYS.WRI$_OPTSTAT_HISTGRM_HISTORY: ADDED INTERVAL PARTITION SYS_P224 (43903) VALUES LESS THAN (TO_DATE(' 2020-03-15 00:00:00', 'SYYYY-MM-DD HH24:MI:SS', 'NLS_CALENDAR=GREGORIAN'))
2020-03-14T13:37:05.255555+00:00
XEPDB1(3):TABLE SYS.WRI$_OPTSTAT_HISTHEAD_HISTORY: ADDED INTERVAL PARTITION SYS_P241 (43903) VALUES LESS THAN (TO_DATE(' 2020-03-15 00:00:00', 'SYYYY-MM-DD HH24:MI:SS', 'NLS_CALENDAR=GREGORIAN'))
XEPDB1(3):TABLE SYS.WRI$_OPTSTAT_HISTGRM_HISTORY: ADDED INTERVAL PARTITION SYS_P244 (43903) VALUES LESS THAN (TO_DATE(' 2020-03-15 00:00:00', 'SYYYY-MM-DD HH24:MI:SS', 'NLS_CALENDAR=GREGORIAN'))
2020-03-14T13:37:37.974180+00:00
XEPDB1(3):Resize operation completed for file# 9, old size 256000K, new size 266240K
2020-03-14T13:38:00.841460+00:00
Resize operation completed for file# 1, old size 849920K, new size 860160K
2020-03-14T13:38:32.483165+00:00
Errors in file /opt/oracle/diag/rdbms/xe/XE/trace/XE_j003_360.trc:
ORA-12012: error on auto execute of job "SYS"."ORA$AT_OS_OPT_SY_1"
ORA-20001: Statistics Advisor: Invalid task name for the current user
ORA-06512: at "SYS.DBMS_STATS", line 49540
ORA-06512: at "SYS.DBMS_STATS_ADVISOR", line 881
ORA-06512: at "SYS.DBMS_STATS_INTERNAL", line 21631
ORA-06512: at "SYS.DBMS_STATS_INTERNAL", line 23763
ORA-06512: at "SYS.DBMS_STATS", line 49528

**Comment by KamilJedrzejuk** [comment]
Chceck this walk around https://github.com/KamilJedrzejuk/oracle18.4.0xe-docker

**Comment by Djelibeybi** [comment]
> Chceck this walk around https://github.com/KamilJedrzejuk/oracle18.4.0xe-docker

Why not sign the Oracle Contributor Agreement and submit that as an additional Dockerfile for this repo?

**Comment by KamilJedrzejuk** [comment]
@Djelibeybi this repo had same issues so IT was deleted.
You can check this https://github.com/KamilJedrzejuk/oracle18c-xe
But is not base on official

**Comment by KamilJedrzejuk** [comment]
Why image with Oracle18c is several times bigger than PostgreSQL? Can You Explain me, I think that Oracle is not the best choice for development, I want to use a Testcontainers library https://programmerfriend.com/spring-boot-integration-testing-done-right/

**Comment by patric-r** [comment]
@gvenzl Has this issue been finally fixed?

**Comment by rafaelosoto** [comment]
This issue has not been fixed and probably stems from how oracle approaches software as a whole. 

In general when developing a docker application you start from the ground up making sure it is quick to start so that it is available for other services(especially if you can store database data elsewhere)

Oracle's approach is to have an existing slow product and shove it into a container. 

@KamilJedrzejuk thank you so much will give this a shot. Not sure why this issue is closed it is one of the biggest issues in this repo.

**Comment by KamilJedrzejuk** [comment]
Thanks for explanation.

**Comment by mh5** [comment]
I found a way to move away the slowness from the first container startup to the image build. In the Dockerfile, I added the below line, right before the `CMD` line.

    RUN $ORACLE_BASE/$RUN_FILE --nowait

The file `$ORACLE_BASE/$RUN_FILE` is the entrypoint script itself. That is why I found it.

By executing the entrypoint once during the image build, the database runs and the initializes itself. The `--nowait` is a parameter that makes the entrypoint terminate after initialization, I am thankful it exists.
